### PR TITLE
Initial design doc for scheduler.

### LINKED ISSUE #6781: Document how the Kubernetes scheduler works
Write up a short summary of how the Kubernetes scheduler works for those curious about it.

### PR DESCRIPTION
**Pull request by davidopp**
Fixes #6781

@dingh let me know what you think. Please feel free to send a PR to add stuff. This turned out to be a little more than an outline, but I tried to keep it simple.

### CODE REVIEW

**Comment by mikedanese** [comment]
@HaiyangDING

File: docs/design/scheduler.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+(or, if there are multiple such nodes, then one of them is chosen at random). The code
+for this main scheduling loop is in the function `Schedule()` in
+[plugin/pkg/scheduler/generic_scheduler.go](../../plugin/pkg/scheduler/generic_scheduler.go)
+
+## Scheduler extensibility
+
+The scheduler is extensible: the cluster administrator can choose which of the pre-defined
+scheduling policies to apply, and can add new ones. The built-in predicates and priorities are
+defined in [plugin/pkg/scheduler/predicates.go](../../plugin/pkg/scheduler/predicates.go) and
+[plugin/pkg/scheduler/priorities.go](../../plugin/pkg/scheduler/priorities.go), respectively.
```
Comment (mikedanese) [inline docs/design/scheduler.md]: This link and the link on the line above are broken.

File: docs/design/scheduler.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+
+# The Kubernetes Scheduler
+
+The Kubernetes scheduler runs as a process alongside the other master
+components such as the API server. Its interface to the API server is to watch
+for Pods with an empty PodSpec.NodeName, and for each them to post a Binding
```
Comment (derekwaynecarr) [inline docs/design/scheduler.md]: "Its interface to the API server is to watch for Pods with an empty PodSpec.NodeName, and for each pod, it posts a Binding indicating where the Pod should be scheduled."

File: docs/design/scheduler.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+The scheduler tries to find a node for each Pod, one at a time, as it notices
+these Pods via watch. There are three steps. First it applies a set of "predicates" that filter out
+inappropriate nodes. For example, if the PodSpec specifies resource limits, then the scheduler
+will filter out nodes that don't have at least that much resources available (computed
+as the capacity of the node minus the sum of the resource limits of the Pods that
+are already running on the node). Second, it applies a set of "priority functions"
+that rank the nodes that weren't filtered out by the predicate check. For example,
+if the PodSpec specified resource limits, then it favors the least-loaded
+machine (measured in terms of the sum of the limits of the Pods already on each node).
+Finally, the node with the highest priority is chosen
```
Comment (HaiyangDING) [inline docs/design/scheduler.md]: May we should say something about weight?

Each priority function is assigned with a weight (a positive number) representing how much this priority function would affected the final score when ranking each node.

File: docs/design/scheduler.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+The scheduler tries to find a node for each Pod, one at a time, as it notices
+these Pods via watch. There are three steps. First it applies a set of "predicates" that filter out
+inappropriate nodes. For example, if the PodSpec specifies resource limits, then the scheduler
+will filter out nodes that don't have at least that much resources available (computed
+as the capacity of the node minus the sum of the resource limits of the Pods that
+are already running on the node). Second, it applies a set of "priority functions"
+that rank the nodes that weren't filtered out by the predicate check. For example,
+if the PodSpec specified resource limits, then it favors the least-loaded
+machine (measured in terms of the sum of the limits of the Pods already on each node).
```
Comment (HaiyangDING) [inline docs/design/scheduler.md]: The occupancy of each node is measured by the sum of the limits of the Pods already on the node divided by the capacity of the node.

File: docs/design/scheduler.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+file specifying which scheduling policies to use. See
+[examples/scheduler-policy-config.json](../../examples/scheduler-policy-config.json) for an example
+config file. (Note that the config file format is versioned; the API is defined in
+[plugin/pkg/scheduler/api/](../../plugin/pkg/scheduler/api/)).
+Thus to add a new scheduling policy, you should modify predicates.go or priorities.go,
+and either register the policy in `defaultPredicates()` or `defaultPriorities()`, or use a policy config file.
+
+## Exploring the code
+
+If you want to get a global picture of how the scheduler work, you can start in
```
Comment (HaiyangDING) [inline docs/design/scheduler.md]: "If you want to get a global picture of how the scheduler works, ......"

Hi, I think this is a great design doc for scheduler. I have two other comments above, but I think leaving the doc the way it is now is also a good idea.

File: docs/design/scheduler.md (hunk 2 above)
Comment (davidopp) [inline docs/design/scheduler.md]: Done

File: docs/design/scheduler.md (hunk 1 above)
Comment (davidopp) [inline docs/design/scheduler.md]: Thanks, fixed

File: docs/design/scheduler.md (hunk 4 above)
Comment (davidopp) [inline docs/design/scheduler.md]: Good point. I've clarified the wording, also to reflect that we're adding spreading. (Though my intention here was *not* to describe the algorithms in detail -- I think we can have a separate section on that. You are welcome to write that. For this section I just wanted to give a flavor.)

File: docs/design/scheduler.md (hunk 3 above)
Comment (davidopp) [inline docs/design/scheduler.md]: I think we can have a later section where we talk about the algorithms in more detail. Based on @erictune 's comment I wasn't sure how much detail we should get into. So I wanted this PR to be high-level, talking only about things that are unlikely to change soon.

**Comment by davidopp** [comment]
Addressed the comments, PTAL.

@HaiyangDING Please feel free to follow up with a PR that goes into more detail. I just wanted to make sure we got something simple and high-level in first. (And that would be non-controversial relative to @erictune 's comment :) )

File: docs/design/scheduler.md (hunk 3 above)
Comment (HaiyangDING) [inline docs/design/scheduler.md]: I totally agree with you on this. The conception of weight itself is not hard to understand but introducing it here is likely to make this doc a little messy.

**Comment by HaiyangDING** [comment]
@davidopp 

I think it is a great doc, simple and clear. And thank you for sharing your thoughts about this PR:)

I will send a PR to go through the details of the current algorithm and keep updating it once accepted.  The PR will be delivered in a couple of days and by then, it would be really nice if you could give me some advice to improve it.

File: docs/design/scheduler.md (hunk 4 above)
Comment (HaiyangDING) [inline docs/design/scheduler.md]: The updated description is good enough, simple and clear. I see your points here, and I will try to work out another section for more details like I have said in the comment below.

**Comment by mikedanese** [comment]
@davidopp you need to run ./hack/run-gendoc.sh to add the analytics footer.

**Comment by mikedanese** [comment]
This is LGTM after you add the analytics footer ^^^

**Comment by mikedanese** [comment]
This is LGTM after you add the analytics footer ^^^

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by davidopp** [comment]
Thanks, can't believe I forgot that right after I asked about it!

I've run gendocs and squashed the commits. PTAL.

**Comment by mikedanese** [comment]
Will add ok-to-merge on shippable green. cc @erictune

**Comment by davidopp** [comment]
@k8s-bot ok to test

**Comment by Dieterbe** [comment]
@davidopp thanks for your write-up.
for anyone else trying to find the documentation now and not finding it anywhere (the file contributed here was removed, the kubernetes website also doesn't have it for some reason), it lives at:
https://github.com/kubernetes/community/blob/master/contributors/devel/scheduler_algorithm.md
https://github.com/kubernetes/community/blob/master/contributors/devel/scheduler.md

**Pull request merged**